text/0000-poll-drop-ready.md

```diff
@@ -0,0 +1,480 @@
+
+## Guarantees about when `poll_drop_ready` is called
+
+In general, users cannot assume that `poll_drop_ready` will ever be called,
+just as they cannot assume that destructors will run at all. However, they can
+assume less about `poll_drop_ready` than they can about `drop`, and that should
+be noted.
+
+In particular, users cannot safely assume that their values will be dropped in
+an async context, regardless of how they are intended to be used. It's this
+easy to supress the `poll_drop_ready` call of a value:
+
+```rust
+async fn ignore_poll_drop_ready<T>(x: T) {
+    // by passing `x` to `drop`, a non-async context, `poll_drop_ready` is
+    // never called when dropping `x`.
+    drop(x);
+}
+```
+
+However, we do guarantee that values dropped in an async context *will* have
+`poll_drop_ready` called, and we even guarantee the drop order between
+variables and between fields of a type. When the user sees a value go out of
+scope *in an async context*, they know that `poll_drop_ready` is called. And we
+do guarantee that the destructors of all fields of that type will be called.
+
+## Changes to destructor calls in async contexts
+
+In async contexts, we change the generated destructor calls to insert, before
+each call, this code:
+
+```rust
+while let Poll::Pending = mem::poll_drop_ready(obj, cx) {
+    yield Poll::Pending;
+}
+```
+
+This "prepares" the object to be destroyed by calling its destructor, calling
+`poll_drop_ready` on the object and all of its fields. This is called before
+any destructors are run on that object.
+
+In non-async contexts, no change is made to destructors. This means that users
+writing low level primitives which may want to run async destructor code on
+values they are destroying (such as inside the poll method of a
+manually-implemented future) will need to call `mem::poll_drop_ready`
+themselves, or this code will be skipped.
+
+Therefore, there are two cases in which users will have to take special care to
+ensure that async destructors are called:
+
+1. If they would already have to call `drop_in_place` to ensure that normal
+   destructors are called (as in data structures like `Vec`).
+2. If they are dropping a value that may have an async destructor inside a
+   poll method.
+
+## Drop glue in `mem::poll_drop_ready`
+
+`mem::poll_drop_ready` will first call `Drop::poll_drop_ready` on this value,
+before recursively calling `mem::poll_drop_ready` on every field of the value.
+It will logically "AND" all of the return values, so that it will return
+`Poll::Ready` only if all calls return `Poll::Ready`, and otherwise return
+`Poll::Pending`. In pseudo-code:
+
+```rust
+let mut ready = self.poll_drop_ready(cx);
+$(
+    ready &= mem::poll_drop_ready(&mut self.fieldN, cx);
+)*
+ready
+```
+
+### The necessity of fused semantics
+
+Calls to `poll_drop_ready` occur in a loop, until every recursive subcall
+returns `Ready`. These calls are (by necessity) stateless: we would be required
```

Comment:
I wouldn't call this stateless, since fields are clearly stateful - their state is modified by polling. I guess you mean the there is no state to record which fields are already done and that calling the function needs to be idempotent - but this is not what I first think about when reading `stateless`